### PR TITLE
Publish the commands sent to IMC after limits

### PR DESCRIPTION
**Pull request by JitskedeVries**
In combination with this pullrequest: https://github.com/project-march/march-iv/pull/288, the commands that are adjusted by the limitinterface are published onto a topic. 

More explanation about why we want to publish this can be found in that pull request.

### CODE REVIEW

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 1]
```diff
@@ -249,18 +253,37 @@ void MarchHardwareInterface::write(ros::Duration elapsed_time)
 {
   positionJointSoftLimitsInterface.enforceLimits(elapsed_time);
 
+  ROS_DEBUG("Before limits: Trying to actuate joint %s, to %lf "
```
Comment (sopje) [inline march_hardware_interface/src/march_hardware_interface.cpp]: don't you want this to loop through all the joints like the after limits ROS_debug message?

**Review by sopje — COMMENTED** [review]
It worked on the rotational test joint, please take a look at my small comment

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 2]
```diff
@@ -247,20 +251,43 @@ void MarchHardwareInterface::read(ros::Duration elapsed_time)
...
     {
       ROS_DEBUG("After limits: Trying to actuate joint %s, to %lf rad, %f "
                 "speed, %f effort.",
                 joint_names_[i].c_str(), joint_position_command_[i], joint_velocity_command_[i],
                 joint_effort_command_[i]);
       marchRobot.getJoint(joint_names_[i]).actuateRad(static_cast<float>(joint_position_command_[i]));
+
+      after_limit_joint_command_pub_->msg_.name.push_back(singleJoint.getName());
+      after_limit_joint_command_pub_->msg_.position_command.push_back(joint_position_command_[i]);
+      after_limit_joint_command_pub_->msg_.effort_command.push_back(0);
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: This should be moved to another method instead of cluttering the read method

**Review by Ishadijcks — CHANGES_REQUESTED** [review]

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 3]
```diff
@@ -38,9 +38,13 @@ MarchHardwareInterface::~MarchHardwareInterface()
 void MarchHardwareInterface::init()
 {
   // Initialize realtime publisher for the IMotionCube states
-  imc_state_pub_ = RtPublisherPtr(
+  imc_state_pub_ = RtPublisherImcStatePtr(
       new realtime_tools::RealtimePublisher<march_shared_resources::ImcErrorState>(this->nh_, "/march/imc_states/", 4));
 
+  after_limit_joint_command_pub_ = RtPublisherAfterLimitJointCommandPtr(
+      new realtime_tools::RealtimePublisher<march_shared_resources::AfterLimitJointCommand>(
+          this->nh_, "/march/after_limit_joint_command/", 4));
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp:46]: Please pick a topic closer related to the controller.

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 4]
```diff
@@ -251,198 +255,236 @@ void MarchHardwareInterface::read(ros::Duration elapsed_time)
...
+        march4cpp::Joint singleJoint = marchRobot.getJoint(joint_names_[i]);
+
+        if (marchRobot.getJoint(joint_names_[i]).canActuate())
+        {
+            ROS_DEBUG("After limits: Trying to actuate joint %s, to %lf rad, %f "
+                      "speed, %f effort.",
+                      joint_names_[i].c_str(), joint_position_command_[i], joint_velocity_command_[i],
+                      joint_effort_command_[i]);
+            marchRobot.getJoint(joint_names_[i]).actuateRad(static_cast<float>(joint_position_command_[i]));
+        }
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp:279]: Please refactor this to use the style develop uses at the moment.
```
march4cpp::Joint joint = marchRobot.getJoint(joint_names_[i]);
```
Once and use the variable later

**Review by Ishadijcks — CHANGES_REQUESTED** [review]
I don't understand why there are so many changes in this pr.
It looks like many refactors have been merge since you've branched from develop. Can you either merge develop into it and clean it up, create a new branch from the current develop with your changes

**Comment by JitskedeVries** [comment]
> I don't understand why there are so many changes in this pr.
> It looks like many refactors have been merge since you've branched from develop. Can you either merge develop into it and clean it up, create a new branch from the current develop with your changes

It is kind of weird indeed. Will do this!

**Pull request closed without merging**